packages/material-ui/src/Menu/Menu.js

```diff
@@ -115,13 +176,65 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   });
 
+  const handleSetOpenSubMenuIndex = (value) => {
+    if (value === null) {
+      if (contentAnchorRef.current.parentElement) {
+        contentAnchorRef.current.parentElement.children[openSubMenuIndex].focus();
+      }
+    }
+    setOpenSubMenuIndex(value);
+  };
+
   const items = React.Children.map(children, (child, index) => {
+    if (!React.isValidElement(child)) {
+      return undefined;
+    }
+
+    const { subMenu, onMouseMove: onMouseMoveChildProp } = child.props;
+    const { anchorEl } = other;
+
+    const hasSubMenu = Boolean(subMenu);
+    const parentMenuOpen = Boolean(anchorEl);
+
+    const additionalProps = {};
+
+    // This is the original purpose of this React.Children.map and is basically unchanged.
     if (index === activeItemIndex) {
+      additionalProps.ref = (instance) => {
+        contentAnchorRef.current = instance;
+        setRef(child.ref, instance);
+      };
+    }
+
+    // If the current Menu item in this map has a subMenu,
+    // we need the parent Menu to orchestrate its subMenu
+    if (hasSubMenu && parentMenuOpen) {
+      additionalProps.onArrowRightKeydown = (e) => {
+        if (e.key === 'ArrowRight') {
+          e.preventDefault();
+          setOpenSubMenuIndex(index);
+        }
+      };
+      additionalProps.openSubMenu = index === openSubMenuIndex && !entering;
+      additionalProps.setParentOpenSubMenuIndex = handleSetOpenSubMenuIndex;
+    }
+
+    // If there are ANY children with subMenus, then ALL
+    // of the children need to know how to close any open subMenus
+    // and reset the state that controls which subMenu is open.
+    if (atLeastOneSubMenu) {
+      additionalProps.onMouseMove = (e) => {
+        setOpenSubMenuIndex(index);
+        if (onMouseMoveChildProp) {
+          onMouseMoveChildProp(e);
+        }
+      };
+      additionalProps.onParentClose = handleOnClose;
+    }
+
+    if (Object.keys(additionalProps).length > 0) {
       return React.cloneElement(child, {
-        ref: (instance) => {
-          contentAnchorRef.current = instance;
-          setRef(child.ref, instance);
-        },
+        ...additionalProps,
```

Comment:
nit: seems like we don't need to spread here.

Reply:
Apologies, I'm realizing I missed several of your responses that were buried in collapsed parts of the conversation and I haven't been getting emails for all of these, it seems.

I'll take a look at this once I get past the latest issues I seem to have introduced during merge conflict resolution for the rebase on the emotion changes.